tools/ci_build/exclude_unused_ops.py

```diff
@@ -150,9 +161,44 @@ def _exclude_unused_ops_in_provider(operators, provider_path):
     onnx_versioned_typed_op_len = len(onnx_versioned_typed_op)
     end_marks = tuple([');', ')>', ')>,', ')>,};', ')>};'])
 
-    def should_exclude_op(domain, op_type, opset_from, opset_to=None):
+    # we validate that all the operator registrations are correct whilst running this script.
+    # all operator entries are ordered by opset, so we can just track the most recent entry
+    # for each domain + operator type.
+    # we should only see versioned entries until the last entry and the last entry must always be unversioned.
+    # the end version of the previous entry must always be one less than the start version of the next entry.
+    last_op_registrations = {}
+
+    def validate_registration(domain, op_type, opset_from_str, opset_to_str):
+        key = domain + ':' + op_type
+        prev_from, prev_to = last_op_registrations[key] if key in last_op_registrations else (None, None)
+        opset_from = int(opset_from_str)
+        opset_to = int(opset_to_str) if opset_to_str else None
+
+        if prev_from:
+            # a typed registration where the to/from matches for each entry so nothing to update
+            if prev_from == opset_from and prev_to == opset_to:
+                return
+
+            # previous registration was unversioned but should have been
+            if not prev_to:
+                log.error("Invalid registration for {}. Registration for opset {} has no end version."
+                          .format(key, prev_from))
+                sys.exit(-1)
+
+            # previous registration end opset is not adjacent to the start of the next registration
+            if prev_to != opset_from - 1:
+                log.error("Invalid registration for {}. Registration for opset {} should have end version of {}"
+                          .format(key, prev_from, opset_from - 1))
+                sys.exit(-1)
+
+        last_op_registrations[key] = (opset_from, opset_to)
+
+    def should_exclude_op(domain, op_type, opset_from, opset_to, is_registration):
         '''check if should exclude the op from build'''
 
+        if is_registration:
+            validate_registration(domain, op_type, opset_from, opset_to)
```

Comment:
>validate_registration [](start = 12, length = 21)

How about call it independently before should_exclude_op since the purposes are differed - show_exclude_op should consider solely on inputted arguments like '--model_path' or '--config_path', while validate_registration hinges on deprecated list.

Reply:
Refactored.

---
In reply to: [523941722](https://github.com/microsoft/onnxruntime/pull/5817#discussion_r523941722) [](ancestors = 523941722)